Memory profiling and fixing memory leak

angelala3252 opened the pull request:
<!-- Thanks for submitting a pull request! Please provide information so we can quickly review your pull request. -->

**What kind of change does this PR introduce?**
Bugfix and add memory profiling.

<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Checklist**
* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] If applicable, I have added tests to prove my fix is functional/effective
* [x] I have linted my code locally prior to submission
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation (e.g. README.md, requirements.txt)
* [x] New and existing unit tests pass locally with my changes

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? Try to link to an open issue. -->

This PR responds to [Issue #5](https://github.com/MLDSAI/OpenAdapt/issues/5). The goal is to add memory profiling to `record.py`, add memory usage to performance plots, and fix the memory leak caused by the screenshots described [here](https://github.com/MLDSAI/OpenAdapt/issues/5#issuecomment-1593323315). 

**How can your code be run and tested?**

```
git remote add angelala3252 https://github.com/angelala3252/OpenAdapt.git
git fetch angelala3252
git checkout feat/memory
pip install -r requirements.txt
python -m openadapt.record "testing out openadapt"
```

<!-- See the README.md for examples. -->

KrishPatel13 commented:
@angelala3252
Resolved Merge Conflicts. Let me know if there are any issues :)

abrichr commented on openadapt/record.py:
```diff
@@ -618,12 +634,30 @@ def record(
+                stat.size_diff / 1024,
+                stat.size / 1024,
+                stat.count_diff,
+                stat.count))
```
What do you think about:
```
        new_KiB = stat.size_diff / 1024
        total_KiB = stat.size / 1024
        new_blocks = stat.count_diff
        total_blocks = stat.count
        logger.info(f"{new_KiB=} {total_KiB=} {new_blocks=} {total_blocks=}")
```

angelala3252 replied:
```
    for stat in stats[:3]:
        new_KiB = stat.size_diff / 1024
        total_KiB = stat.size / 1024
        new_blocks = stat.count_diff
        total_blocks = stat.count
        source = stat.traceback.format()[0]
        logger.info(f"{new_KiB=} {total_KiB=} {new_blocks=} {total_blocks=} {source=}")
```

output:
![image](https://github.com/MLDSAI/OpenAdapt/assets/88949118/da1f7493-bd70-4dd8-b4e6-a5d676c23e31)

abrichr commented on openadapt/record.py:
```diff
@@ -618,12 +634,30 @@ def record(
+        for line in stat.traceback.format():
+            print(line)
+
+    tr.print_diff()
```
Is there a way to get the value as a string, e.g.:
```
trace_str = tr.get_diff_str()
logger.info(f"trace_str=\n{trace_str}")
```

angelala3252 replied:
Yes, this is the output now:
![image](https://github.com/MLDSAI/OpenAdapt/assets/88949118/ec43bf49-8ec1-463d-a0ed-cce37e5912bd)

angelala3252 commented:
Updates: 

- the memory leak seems to have been resolved after adding `del prev_event` to `process_events`. sleeping for any amount of time in `read_screen_events` had no effect. 

Comparison of memory usage before and after any changes on a shorter recording:
tracemalloc found 202mb from screenshots, and pympler found 189.85 MB
Performance graph:
![performance-before-changes](https://github.com/MLDSAI/OpenAdapt/assets/88949118/6e77fbe2-2e17-403e-bd69-081c02967d02)

After after adding `del prev_event`:
tracemalloc found 16mb from screenshots, and pympler found 7.91 MB
Performance graph:
![performance-del](https://github.com/MLDSAI/OpenAdapt/assets/88949118/fcbe96dc-7ab8-459c-9944-48352fa37d3d)

This is a significant decrease in memory usage, and I also tested it with longer recordings and got great results. Before, longer recordings would cause Pycharm to take up to 4-5000 MB of memory when a lot of input was given, and around 2-3000 MB when no input was given. After this change, Pycharm only takes up ~1200MB on longer recordings and stays pretty constant even if a lot of input is given.

abrichr commented:
Excellent work @angelala3252 ! What else is required here in order to be ready for review?

abrichr commented on requirements.txt:
```diff
@@ -34,6 +34,7 @@ sumy==0.11.0
 transformers==4.29.2
-python-dotenv==1.0.0
+pympler==1.0.1
+nicegui==1.2.16
```
Can we please keep this alphabetical?

abrichr commented on openadapt/record.py:
```diff
@@ -24,6 +24,14 @@
+
+tracemalloc.start(25)
+
+snapshots = []
```
What do you think about naming this e.g. `performance_snapshots` or similar?

abrichr commented on openadapt/record.py:
```diff
@@ -24,6 +24,14 @@
+
+import tracemalloc
+
+tracemalloc.start(25)
```
Where does the number 25 come from? Can you please name this in a constant?

abrichr replied:
25 is fine just please define it with a self documenting constant

angelala3252 commented:
@abrichr The only thing left to do is plotting the memory usage over time

abrichr commented on openadapt/record.py:
```diff
@@ -32,12 +40,15 @@
 }
-PLOT_PERFORMANCE = False
-
+PLOT_PERFORMANCE = True
```
What do you think about moving this to `config.py`?

angelala3252 replied:
done in [a9828dd](https://github.com/MLDSAI/OpenAdapt/pull/270/commits/a9828dd9dd5de9e7431e6ba713fa1169e1b8b2e1)

abrichr commented on openadapt/record.py:
```diff
@@ -618,12 +631,29 @@ def record(
 
+    collect_stats()
+
+    stats = performance_snapshots[-1].compare_to(performance_snapshots[-2], 'lineno')
```
Can you please clarify why we're only comparing the second-to-last to the last? Is it because the number of expected snapshots is two? In that case, what do you think about something like this:

```
assert len(performance_snapshots) == 2, performance_snapshots
first_snapshot, last_snapshot = performance_snapshots
...
```

angelala3252 replied:
The expected number is two. I've added your code in [e96b942](https://github.com/MLDSAI/OpenAdapt/pull/270/commits/e96b9420bd9ad3bdf15e2f7561cfea451eb06797)

abrichr commented on openadapt/record.py:
```diff
@@ -618,12 +631,29 @@ def record(
+
+    stats = performance_snapshots[-1].compare_to(performance_snapshots[-2], 'lineno')
+
+    for stat in stats[:3]:
```
Why do we start at stats[:3]?

angelala3252 replied:
just `stats` would show every single memory allocation and there are a lot, so it fills up the logging with a lot of insignificant memory allocations that only take a few bytes. I've defined a constant `NUM_MEMORY_STATS` for the 3, so we could change that to 5 or 10 or however many stats we want to see.

angelala3252 commented:
After testing recording a bit more, I don't think that the memory leak has actually been fixed, because it's very inconsistent. Sometimes, the memory usage returns to the previous memory usage before the changes, using 4-5 GB of memory, and sometimes the memory stays around 600 MB. I'm not sure how the results I got previously were so good, but perhaps it was because the recordings were relatively shorter and didn't have as many ActionEvents, but the memory usage doesn't stay as low when I produce more ActionEvents by typing really fast. The problem is still the screenshots, and it seems like the screenshots are just being put into the queue faster than they are processed because eventually, the memory usage goes back down to a base level of around 400MB. I will have to continue looking into the memory leak and how to fix this.

abrichr commented:
Thanks @angelala3252 ! As discussed this is expected behavior and we can delay addressing this for now.

angelala3252 commented:
- added memory to the legend in [2918fc6](https://github.com/MLDSAI/OpenAdapt/pull/270/commits/2918fc6c9a066d9fc616bb512464780bf6eaefe8)
- updated code to write memory data to database instead of a file in [3a8d4b3](https://github.com/MLDSAI/OpenAdapt/pull/270/commits/3a8d4b3d527c387a056f3987b55c4e2a98d9400b)

abrichr commented:
Thanks @angelala3252 ! Can you please show an updated screenshot? Also please fix merge conflicts 🙏

angelala3252 commented:
Updated screenshot:
![legend2](https://github.com/MLDSAI/OpenAdapt/assets/88949118/c56d01a3-2808-45b4-8013-bfccebbc21c3)

angelala3252 commented:
@abrichr resolved merge conflicts and addressed comments

abrichr commented on openadapt/record.py:
```diff
@@ -34,12 +36,34 @@
 }
-PLOT_PERFORMANCE = False
+PLOT_PERFORMANCE = config.PLOT_PERFORMANCE
+BYTE_TO_MB = float(2**20)
```
@angelala3252 thank you for naming this constant. However it's still not clear to me what it represents. There are 10**6 bytes in a megabyte. Can you please clarify?

abrichr replied:
I had a brain malfunction 😅

abrichr commented on openadapt/record.py:
```diff
@@ -34,12 +36,34 @@
-PLOT_PERFORMANCE = False
+PLOT_PERFORMANCE = config.PLOT_PERFORMANCE
+BYTE_TO_MB = float(2**20)
+NUM_MEMORY_STATS = 3
```
Is this just to limit the log output?

angelala3252 replied:
Yeah, without it every single memory allocation would be displayed, even the tiny insignificant ones. If you want more output I could change it to something like 5 or 10.

abrichr commented on openadapt/utils.py:
```diff
@@ -439,13 +438,15 @@ def plot_performance(recording_timestamp: float = None) -> None:
     perf_stats = crud.get_perf_stats(recording_timestamp)
     perf_stat_dicts = rows2dicts(perf_stats)
+    mem_stats = crud.get_memory_stats(recording_timestamp)
+    mem_stat_dicts = rows2dicts(mem_stats)
```
Is this necessary? Why not just access `mem_stats` directly?

angelala3252 replied:
fixed in [35e9de1](https://github.com/MLDSAI/OpenAdapt/pull/270/commits/35e9de1cb44a9ec16dc36963019c6aed55a1ade2)

abrichr commented on requirements.txt:
```diff
@@ -13,6 +13,7 @@ loguru==0.6.0
 matplotlib==3.6.2
 moviepy==1.0.3
 mss==6.1.0
+nicegui==1.2.16
```
Is this necessary?

abrichr commented on requirements.txt:
```diff
@@ -21,6 +22,9 @@ pyinstaller
 presidio_image_redactor==0.0.46
+pympler==1.0.1
+python-dotenv==1.0.0
+python-Levenshtein==0.21.1
```
Is this necessary?

abrichr commented on requirements.txt:
```diff
@@ -21,6 +22,9 @@ pyinstaller
 presidio_anonymizer==2.2.32
 presidio_image_redactor==0.0.46
+pympler==1.0.1
+python-dotenv==1.0.0
```
Is this necessary?

angelala3252 commented:
@abrichr ready for review

abrichr commented on openadapt/models.py:
```diff
@@ -314,3 +314,12 @@ class PerformanceStat(db.Base):
+
+    id = sa.Column(sa.Integer, primary_key=True)
+    recording_timestamp = sa.Column(sa.Integer)
+    memory_usage = sa.Column(ForceFloat)
```
@angelala3252 can you please rename this to something more descriptive, e.g. `memory_usage_kb` (or similar)?

Also, I noticed in the migration that the type does not match:

```
                    sa.Column('memory_usage', sa.Integer(), nullable=True),
                    sa.Column('timestamp', sa.Integer(), nullable=True),
```

And when I try to run `alembic upgrade head` I get:

```
KeyError: '3d3239ae4849'
```

This suggests to me that you created the revision off of a stale database state.

Can you please:

- delete openadapt.db
- check out main
- `alembic upgrade head`
- check out this branch
- `alembic revision ...`
- replace the current migration script with the new one

You may need to import the ForceFloat type.

angelala3252 replied:
done in [fb290bb](https://github.com/MLDSAI/OpenAdapt/pull/270/commits/fb290bbadb2ee7d237611e61495a9497da5868e1) and [0d0c842](https://github.com/MLDSAI/OpenAdapt/pull/270/commits/0d0c8426b5fa5cfa57cadb0870e2133fac239683)

abrichr commented:
Thanks @angelala3252 ! I get this exception when creating a recording:

```

2023-07-02 10:45:43.902 | INFO     | openadapt.config:<module>:130 - DIRNAME_PERFORMANCE_PLOTS=performance
2023-07-02 10:45:45.232 | INFO     | __main__:wrapper_logging:88 -  -> Enter: record(foo)
2023-07-02 10:45:45.252 | INFO     | __main__:record:614 - task_description='foo'
2023-07-02 10:45:45.253 | INFO     | __main__:wrapper_logging:88 -  -> Enter: create_recording(foo)
2023-07-02 10:45:45.310 | INFO     | __main__:create_recording:555 - recording=Recording(id=5, timestamp=1688309145.2541811, monitor_width=1512, monitor_height=982, double_click_interval_seconds=0.5, double_click_distance_pixels=5, platform='darwin', task_description='foo')
2023-07-02 10:45:45.310 | INFO     | __main__:wrapper_logging:92 -  <- Leave: create_recording(Recording(id=5, timestamp=1688309145.2541811, monitor_width=1512, monitor_height=982, double_click_interval_seconds=0.5, double_click_distance_pixels=5, platform='darwin', task_description='foo'))
2023-07-02 10:45:45.312 | INFO     | __main__:wrapper_logging:88 -  -> Enter: read_window_events(<queue.Queue object at 0x142efb9d0>, <multiprocessing.synchronize.Event object at 0x142fc0280>, 1688309145.2541811)
2023-07-02 10:45:45.318 | INFO     | __main__:read_window_events:429 - starting
2023-07-02 10:45:45.329 | INFO     | __main__:process_events:132 - starting
2023-07-02 10:45:45.335 | INFO     | __main__:read_screen_events:402 - starting
2023-07-02 10:45:45.335 | INFO     | __main__:read_screen_events:402 - starting
2023-07-02 10:45:45.538 | INFO     | __main__:read_window_events:448 - _window_data={'title': 'Terminal OpenAdapt — tmux — 214×51', 'left': 0, 'top': 38, 'width': 1512, 'height': 944, 'window_id': 204051}
2023-07-02 10:45:45.538 | INFO     | __main__:read_window_events:448 - _window_data={'title': 'Terminal OpenAdapt — tmux — 214×51', 'left': 0, 'top': 38, 'width': 1512, 'height': 944, 'window_id': 204051}
2023-07-02 10:45:47.155 | INFO     | __mp_main__:wrapper_logging:88 -  -> Enter: performance_stats_writer(<multiprocessing.queues.Queue object at 0x109ff0790>, 1688309145.2541811, <multiprocessing.synchronize.Event object at 0x143243880>)
2023-07-02 10:45:47.162 | INFO     | __mp_main__:performance_stats_writer:479 - performance stats writer starting
2023-07-02 10:45:47.165 | INFO     | __mp_main__:wrapper_logging:88 -  -> Enter: write_events(action, <function write_action_event at 0x142b91a20>, <multiprocessing.queues.Queue object at 0x1098f4790>, <multiprocessing.queues.Queue object at 0x142baf880>, 1688309145.2541811, <multiprocessing.synchronize.Event object at 0x142bafbb0>)
2023-07-02 10:45:47.172 | INFO     | __mp_main__:write_events:273 - event_type='action' starting
2023-07-02 10:45:47.185 | INFO     | __mp_main__:wrapper_logging:88 -  -> Enter: write_events(screen, <function write_screen_event at 0x144ec1bd0>, <multiprocessing.queues.Queue object at 0x10bbec790>, <multiprocessing.queues.Queue object at 0x144edf880>, 1688309145.2541811, <multiprocessing.synchronize.Event object at 0x144edfbb0>)
2023-07-02 10:45:47.192 | INFO     | __mp_main__:write_events:273 - event_type='screen' starting
2023-07-02 10:45:47.195 | INFO     | __mp_main__:wrapper_logging:88 -  -> Enter: write_events(window, <function write_window_event at 0x1429d1c60>, <multiprocessing.queues.Queue object at 0x109750790>, <multiprocessing.queues.Queue object at 0x1429ef880>, 1688309145.2541811, <multiprocessing.synchronize.Event object at 0x1429efbb0>)
2023-07-02 10:45:47.194 | INFO     | __mp_main__:memory_writer:501 - Memory writer starting
2023-07-02 10:45:47.202 | INFO     | __mp_main__:write_events:273 - event_type='window' starting
Process Process-5:
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/local/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/abrichr/oa/OpenAdapt/openadapt/record.py", line 517, in memory_writer
    crud.insert_memory_stat(
  File "/Users/abrichr/oa/OpenAdapt/openadapt/crud.py", line 118, in insert_memory_stat
    _insert(memory_stat, MemoryStat, memory_stats)
  File "/Users/abrichr/oa/OpenAdapt/openadapt/crud.py", line 39, in _insert
    assert not event_data, event_data
AssertionError: {'memory_usage': 889.0}
```

Any ideas?

abrichr commented on openadapt/record.py:
```diff
@@ -467,6 +493,35 @@ def performance_stats_writer (
+    process = psutil.Process(record_pid)
+
+    while not terminate_event.is_set():
+        mem_usage = getattr(process, "memory_info")()[0] / BYTE_TO_MB
```
@angelala3252 can you please clarify why it is necessary to use `getattr` here instead of just `process.memory_info`?

abrichr commented on openadapt/utils.py:
```diff
@@ -475,6 +475,28 @@ def plot_performance(recording_timestamp: float = None) -> None:
+    timestamps = []
+    mem_usages = []
+    for mem_stat in mem_stats:
+        mem_usages.append(mem_stat.memory_usage)
```
```
2023-07-02 11:06:12.633 | INFO     | __mp_main__:memory_writer:525 - Memory writer done
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/abrichr/oa/OpenAdapt/openadapt/record.py", line 764, in <module>
    fire.Fire(record)
  File "/Users/abrichr/Library/Caches/pypoetry/virtualenvs/openadapt-VBXg4jpm-py3.10/lib/python3.10/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/Users/abrichr/Library/Caches/pypoetry/virtualenvs/openadapt-VBXg4jpm-py3.10/lib/python3.10/site-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/Users/abrichr/Library/Caches/pypoetry/virtualenvs/openadapt-VBXg4jpm-py3.10/lib/python3.10/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/Users/abrichr/oa/OpenAdapt/openadapt/record.py", line 90, in wrapper_logging
    result = func(*args, **kwargs)
  File "/Users/abrichr/oa/OpenAdapt/openadapt/record.py", line 753, in record
    utils.plot_performance(recording_timestamp)
  File "/Users/abrichr/oa/OpenAdapt/openadapt/utils.py", line 482, in plot_performance
    mem_usages.append(mem_stat.memory_usage)
AttributeError: 'MemoryStat' object has no attribute 'memory_usage'
```
